lib/std/Build/CompileStep.zig

```diff
@@ -1053,6 +1053,50 @@ fn addRecursiveBuildDeps(cs: *CompileStep, module: *Module, done: *std.AutoHashM
     if (done.contains(module)) return;
     try done.put(module, {});
     module.source_file.addStepDependencies(&cs.step);
+
+    for (module.include_dirs.items) |include_dir| {
+        cs.include_dirs.append(include_dir) catch @panic("OOM");
+    }
+
+    for (module.lib_paths.items) |lib_path| {
+        cs.addLibraryPath(lib_path);
+    }
+
+    for (module.framework_dirs.items) |framework_dir| {
+        cs.addFrameworkPath(framework_dir);
+    }
+
+    for (module.rpaths.items) |rpath| {
+        cs.addRPath(rpath);
+    }
+
+    for (module.config_headers.items) |config_header| {
+        cs.addConfigHeader(config_header);
+    }
+
+    for (module.libs.items) |lib| {
+        cs.linkLibrary(lib);
+    }
+
+    for (module.frameworks.keys(), module.frameworks.values()) |framework, link_options| {
+        cs.frameworks.put(cs.step.owner.dupe(framework), link_options) catch @panic("OOM");
+    }
+
+    for (module.system_libs.items) |system_lib| {
+        cs.link_objects.append(.{
+            .system_lib = system_lib,
+        }) catch @panic("OOM");
+    }
+
+    for (module.installed_headers.items) |installed_header| {
+        switch (installed_header) {
+            .header => |header| cs.installHeader(header.source_path, header.dest_rel_path),
+            .header_dir_step => |header_dir_step| cs.installHeadersDirectoryOptions(header_dir_step),
+            .config_header => |config_header| cs.installConfigHeader(config_header.step, config_header.options),
+            .lib_step => |lib_step| cs.installLibraryHeaders(lib_step),
+        }
+    }
+
```

Comment:
I don't think any state mutations need to happen here. Modules and their information can be traversed during the make phase.